engine/src/eventqueue.cpp

```diff
@@ -1226,68 +1227,92 @@ static void handle_touch(MCStack *p_stack, MCEventTouchPhase p_phase, uint32_t p
+            if (t_touch -> id == p_id)
+                break;
+        
+        /* If there is a touch, fetch the target and delete the touch record if it
+         * is ending or cancelling a touch. */
+        if (t_touch != nil)
+        {
+            t_target = t_touch -> target;
+            
+            // MW-2011-09-05: [[ Bug 9683 ]] Make sure we remove (and delete the touch) here if
+            //   it is 'end' or 'cancelled' so that a cleartouches inside an invoked handler
+            //   doesn't cause a crash.			
+            if (p_phase == kMCEventTouchPhaseEnded || p_phase == kMCEventTouchPhaseCancelled)
+            {
+                if (t_previous_touch == nil)
+                    s_touches = t_touch -> next;
+                else
+                    t_previous_touch -> next = t_touch -> next;
+                
+                delete t_touch;
+            }
+        }
+    }
+    else
+    {
+        t_touch = new (nothrow) MCTouch;
+        t_touch -> next = s_touches;
+        t_touch -> id = p_id;
+        
+        /* If there is a touch, and it was grabbed by a widget and the widget
+         * still exists, then that widget takes any new touches. */
+        if (s_touches != nullptr &&
+            s_touches->is_widget &&
+            s_touches->target.IsValid())
         {
-            MCwidgeteventmanager->event_touch(MCObjectCast<MCWidget>(t_target),
-                                              p_id, p_phase, t_touch_loc.x, t_touch_loc.y);
+            t_target = s_touches->target.Get();
+            t_touch->is_widget = true;
         }
         else
         {
-            switch(p_phase)
+            t_target = p_stack -> getcurcard() -> hittest(t_touch_loc.x, t_touch_loc.y);
+            t_touch->is_widget = false;
+        }
+
+        t_touch -> target = t_target -> GetHandle();
+        
+        s_touches = t_touch;
+    }
+    
+	if (t_target != nil)
+	{
+        // Touches on widgets are handled differently - they grab all touches in
+        // a sequence.
+        if (t_target->gettype() == CT_WIDGET)
+        {
+            if (MCwidgeteventmanager->event_touch(MCObjectCast<MCWidget>(t_target),
+                                                  p_id, p_phase, t_touch_loc.x, t_touch_loc.y))
             {
-                case kMCEventTouchPhaseBegan:
-                    t_target -> message_with_args(MCM_touch_start, p_id);
-                    break;
-                case kMCEventTouchPhaseMoved:
-                    t_target -> message_with_args(MCM_touch_move, p_id, t_touch_loc.x, t_touch_loc.y);
-                    break;
-                case kMCEventTouchPhaseEnded:
-                    t_target -> message_with_args(MCM_touch_end, p_id);
-                    break;
-                case kMCEventTouchPhaseCancelled:
-                    t_target -> message_with_args(MCM_touch_release, p_id);
-                    break;
+                if (s_touches != nullptr)
```

Comment:
@runrevmark Re my comment below perhaps we want:

```
if (t_target->gettype() == CT_WIDGET)
{
              if (MCwidgeteventmanager->event_touch(MCObjectCast<MCWidget>(t_target),
                                                  p_id, p_phase, t_touch_loc.x, t_touch_loc.y))
              {
                      if (s_touches != nullptr)
                      {
                            s_touches->is_widget = true;
                      }
              }
              return;
}
```

Reply:
Hmmm - yes - you are probably right. At the moment we don't do any 'fallthrough' of events from widgets to the host script object - which this would (as you point out) cause.

So, I think what we want to ensure here is that:
- if a widget hit-tests for the first touch in a sequence, then that widget gets *all* touches in that sequence (i.e. only the first touch is hit-tested)
- if a widget does not hit-test for the first touch in a sequence, then LCS gets *all* touches in that sequence

This is related to the issue of mouse messages not percolating for widgets if the widget doesn't handle them. I'll give that some thought - it might be we can use the same model for mouse as we come up with for touch messages. After all, ideally, if a widget is not interested in touch messages and is not going to do anything with them then LCS should probably get them - just as would be the ideal situation with mouse messages.

Reply:
Hmm... I have always thought that it would be good if just like we have a minimal set of properties that the engine handles for a widget (colors, font, rect etc) that a minimal set of messages would be a reasonable thing too. Then if a widget wants to notify the object of a specific event it uses a different message name. Not sure if that ship has sailed though.

Reply:
I think my concerns here may have been partly unfounded because after looking at `MCWidget::hittest` the widget won't be the target unless at least one of it or it's children is hit, however, we do have an issue that `m_touched_widget` can be reset each time there's a `kMCEventTouchPhaseBegan`. We should probably only try and set it if it's `nullptr` (first touch start) I think.

Reply:
One thing is if widgets grab the touch messages like this then you can't do something like:
- create 2 svg icon widgets
- create a behavior to handle touchMove to relocate the target

In the current implementation the second widget will not get any events. Perhaps what we really need is widgets that handle touch get all the touches once they grab them but LCS continues to send messages to the target which is determined for each event. I'm pretty sure that's what I'd like for mouse and key events too...